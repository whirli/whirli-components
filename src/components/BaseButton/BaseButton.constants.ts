export const LoadingStates = {
  DEFAULT: 'default',
  LOADING: 'loading',
  PROCESSING: 'processing',
  COMPLETE: 'complete',
};

export const PropKeys = {
  TYPE: 'type',
  URL: 'url',
  COLOR: 'color',
  SIZE: 'size',
  STATE: 'state',
};

export const PropValues = {
  [PropKeys.TYPE]: ['button', 'submit'],
  [PropKeys.COLOR]: ['default'],
  [PropKeys.SIZE]: ['default'],
  [PropKeys.STATE]: [
    LoadingStates.DEFAULT,
    LoadingStates.LOADING,
    LoadingStates.PROCESSING,
    LoadingStates.COMPLETE,
  ],
};
