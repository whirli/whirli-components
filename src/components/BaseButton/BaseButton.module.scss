@import '@whirli-components/styles/Breakpoints.module.scss';

// default styles
.button {}

// breakpoint styles
$packageBreakpointStyles: (
  size-default: (),
  color-default: ()
);
