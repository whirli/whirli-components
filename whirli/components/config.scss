@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}