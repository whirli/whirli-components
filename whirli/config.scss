@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}