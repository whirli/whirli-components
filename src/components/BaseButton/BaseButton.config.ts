import { PropType } from 'vue';
import * as PropHelpers from '../../helpers/props';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli/BaseButton/BaseButton.config';

// Package config
import { PropKeys, PropValues } from './BaseButton.constants';
// import type { PropColor } from './BaseButton.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  /*[PropKeys.COLOR]: {
    generateBreakpoints: true,
    classes: {
      default: 'color-black',
      ...config.styles[PropKeys.COLOR]?.classes,
    } as Record<PropColor, string>,
  },*/
};

export const ConfigProps: ComponentProps = {
  /*[PropKeys.COLOR]: {
    required: false,
    type: [String, Object] as PropType<PropColor>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.COLOR),
  },*/
  ...config.props,
};
