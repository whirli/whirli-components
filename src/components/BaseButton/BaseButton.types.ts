import { PropKeys, PropValues } from './BaseButton.constants';

const propButtonType = PropValues[PropKeys.TYPE];
const propColor = PropValues[PropKeys.COLOR];
const propSize = PropValues[PropKeys.SIZE];
const propState = PropValues[PropKeys.STATE];

export type PropButtonType = typeof propButtonType[number];
export type PropColor = typeof propColor[number];
export type PropSize = typeof propSize[number];
export type PropState = typeof propState[number];
