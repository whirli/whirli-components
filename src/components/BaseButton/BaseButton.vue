<template>
  <component
    :is="button"
    :type="props.type"
    :url="props.url"
    :class="classes"
    :state="props.state"
    v-bind="$attrs"
  >
    <slot />
    <template #processing>
      <slot name="processing" />
    </template>
    <template #loading>
      <slot name="loading" />
    </template>
  </component>
</template>

<script setup="props" lang="ts">
// Vue
import { defineProps, computed } from 'vue';
import type { ComputedRef, Component } from 'vue';

// components
import BaseButtonDefault from '../BaseButtonDefault/BaseButtonDefault.vue';
import BaseButtonLink from '../BaseButtonLink/BaseButtonLink.vue';

// Styles
import styles from '@whirli-local/BaseButton/BaseButton.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BaseButton.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const classes = [styles.button, ...makeClasses(ComponentStyles, props, styles)];

const button: ComputedRef<Component> = computed((): Component => {
  return props.url !== undefined ? BaseButtonLink : BaseButtonDefault;
});
</script>
