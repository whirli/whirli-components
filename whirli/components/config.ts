import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
