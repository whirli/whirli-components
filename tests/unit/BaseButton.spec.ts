import { mount } from '@vue/test-utils';

import { propReturnsClass } from '../helpers';

import BaseButton from '@/components/BaseButton/BaseButton.vue';
import { PropKeys, LoadingStates } from '@/components/BaseButton/BaseButton.constants';
import { ConfigStyles } from '@/components/BaseButton/BaseButton.config';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const slots = {
  [LoadingStates.DEFAULT]: 'default',
  [LoadingStates.PROCESSING]: 'processing',
  [LoadingStates.LOADING]: 'loading',
  [LoadingStates.COMPLETE]: 'complete',
};

type LoadingStateKey = keyof typeof LoadingStates;

Object.keys(LoadingStates).forEach((loadingKey): void => {
  describe(`BaseButton.vue - Default - State: ${LoadingStates[loadingKey as LoadingStateKey]}`, () => {
    beforeEach(() => {
      wrapper = mount(BaseButton, {
        slots,
        props: {
          state: LoadingStates[loadingKey as LoadingStateKey],
        },
      });
    });

    it('renders the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders the correct element', () => {
      expect(wrapper.element.tagName).toBe('BUTTON');
    });

    it('displays the correct text', () => {
      expect(wrapper.text()).toMatch(slots[LoadingStates[loadingKey as LoadingStateKey]]);
    });

    it('is the correct color', () => {
      expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-default')).toBe(true);
    });

    it('is the correct size', () => {
      expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-default')).toBe(true);
    });
  });
});

describe('BaseButton.vue - Link: Internal', () => {
  beforeEach(() => {
    wrapper = mount(BaseButton, {
      slots,
      props: {
        url: '/internal',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct element', () => {
    expect(wrapper.element.tagName).toBe('ROUTERLINK');
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(slots[LoadingStates.DEFAULT]);
  });

  it('is the correct color', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-default')).toBe(true);
  });

  it('is the correct size', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-default')).toBe(true);
  });
});

describe('BaseButton.vue - Link: External', () => {
  beforeEach(() => {
    wrapper = mount(BaseButton, {
      slots,
      props: {
        url: 'https://www.whirli.com',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct element', () => {
    expect(wrapper.element.tagName).toBe('A');
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(slots[LoadingStates.DEFAULT]);
  });

  it('is the correct color', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-default')).toBe(true);
  });

  it('is the correct size', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-default')).toBe(true);
  });
});
