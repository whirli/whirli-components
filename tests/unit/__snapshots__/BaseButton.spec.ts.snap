// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton.vue - Default - State: complete renders the component 1`] = `
<button
  class=""
  type="button"
>
  
  
  complete
  
  
</button>
`;

exports[`BaseButton.vue - Default - State: default renders the component 1`] = `
<button
  class=""
  type="button"
>
  
  
  default
  
  
</button>
`;

exports[`BaseButton.vue - Default - State: loading renders the component 1`] = `
<button
  class=""
  type="button"
>
  
  
  loading
  
  
</button>
`;

exports[`BaseButton.vue - Default - State: processing renders the component 1`] = `
<button
  class=""
  type="button"
>
  
  
  processing
  
  
</button>
`;

exports[`BaseButton.vue - Link: External renders the component 1`] = `
<a
  class=""
  href="https://www.whirli.com"
  rel="nofollow noopener"
  state="default"
  target="_blank"
  type="button"
>
  
  
  default
  
  
</a>
`;

exports[`BaseButton.vue - Link: Internal renders the component 1`] = `
<routerlink
  class=""
  state="default"
  to="/internal"
  type="button"
>
  
  
  default
  
  
</routerlink>
`;
